liquibase  used setFetchSize,   Hope to support soon

## Feature Request

**For English only**, other languages will not accept.

Please pay attention on issues you submitted, because we maybe need more details. 
If no response anymore and we cannot make decision by current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Is your feature request related to a problem?

Using liquibase and shardingsphere at the project will be problematic because liquibase will used
setFetchSize

liquibase version 3.6.3

sharding-jdbc-spring-boot-starter  version  4.1.1

### Describe the feature you would like.
support setFetchSize

> 感谢您的举报。您能否提供更多信息？
> 
> 请原始，直接且未经任何分析地提供实际行为。这意味着原始输入/输出数据或原始日志，配置...
> Caused by: liquibase.exception.UnexpectedLiquibaseException: liquibase.exception.DatabaseException: java.sql.SQLFeatureNotSupportedException: setFetchSize
	at liquibase.changelog.StandardChangeLogHistoryService.init(StandardChangeLogHistoryService.java:111)
	at liquibase.Liquibase.checkLiquibaseTables(Liquibase.java:1219)
	at liquibase.Liquibase.update(Liquibase.java:193)
	at liquibase.Liquibase.update(Liquibase.java:179)
	at liquibase.integration.spring.SpringLiquibase.performUpdate(SpringLiquibase.java:353)
	at liquibase.integration.spring.SpringLiquibase.afterPropertiesSet(SpringLiquibase.java:305)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1837)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1774)
	... 16 common frames omitted
Caused by: liquibase.exception.DatabaseException: java.sql.SQLFeatureNotSupportedException: setFetchSize
	at liquibase.snapshot.ResultSetCache.get(ResultSetCache.java:101)
	at liquibase.snapshot.JdbcDatabaseSnapshot$CachingDatabaseMetaData.getTables(JdbcDatabaseSnapshot.java:791)
	at liquibase.snapshot.jvm.TableSnapshotGenerator.snapshotObject(TableSnapshotGenerator.java:33)
	at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:66)
	at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
	at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:69)
	at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
	at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:69)
	at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
	at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:69)
	at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
	at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:69)
	at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
	at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:69)
	at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
	at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:69)
	at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
	at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:286)
	at liquibase.snapshot.DatabaseSnapshot.init(DatabaseSnapshot.java:102)
	at liquibase.snapshot.DatabaseSnapshot.<init>(DatabaseSnapshot.java:59)
	at liquibase.snapshot.JdbcDatabaseSnapshot.<init>(JdbcDatabaseSnapshot.java:38)
	at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:217)
	at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:246)
	at liquibase.snapshot.SnapshotGeneratorFactory.getDatabaseChangeLogTable(SnapshotGeneratorFactory.java:254)
	at liquibase.changelog.StandardChangeLogHistoryService.init(StandardChangeLogHistoryService.java:108)
	... 23 common frames omitted
Caused by: java.sql.SQLFeatureNotSupportedException: setFetchSize
	at org.apache.shardingsphere.shardingjdbc.jdbc.unsupported.AbstractUnsupportedDatabaseMetaDataResultSet.setFetchSize(AbstractUnsupportedDatabaseMetaDataResultSet.java:123)
	at liquibase.snapshot.ResultSetCache$ResultSetExtractor.extract(ResultSetCache.java:293)
	at liquibase.snapshot.ResultSetCache$ResultSetExtractor.extract(ResultSetCache.java:288)
	at liquibase.snapshot.JdbcDatabaseSnapshot$CachingDatabaseMetaData$4.fastFetchQuery(JdbcDatabaseSnapshot.java:832)
	at liquibase.snapshot.ResultSetCache$SingleResultSetExtractor.fastFetch(ResultSetCache.java:347)
	at liquibase.snapshot.ResultSetCache.get(ResultSetCache.java:70)
	... 47 common frames omitted

> 您没有填写错误报告模板，所以我必须问您这里是什么发行版本？什么是应用程序（ShardingSphere-JDBC或ShardingSphere-Proxy）？
> sharding-jdbc-spring-boot-starter version 4.1.1
   liquibase version 3.6.3

> 使用setFetchSize时，设置的值是多少？
>  

> 您提到使用setFetchSize时遇到问题，确切的问题是什么？
>  I didn’t use setFetchSize directly. I used setFetchSize internally in liquibase. I get an error when using both liquibase and Sharding Sphere in the project.

> 最后，如果您有发现并要分析，请在单独的部分中进行。

> Thanks for your report. Can you provide more information?
> 
> Please provide the actual behavior originally, rawly and straightforwardly without any analysis. It means raw input/output data, or raw log, config...
> 
Caused by: liquibase.exception.UnexpectedLiquibaseException: liquibase.exception.DatabaseException: java.sql.SQLFeatureNotSupportedException: setFetchSize
  at liquibase.changelog.StandardChangeLogHistoryService.init(StandardChangeLogHistoryService.java:111)
  at liquibase.Liquibase.checkLiquibaseTables(Liquibase.java:1219)
  at liquibase.Liquibase.update(Liquibase.java:193)
  at liquibase.Liquibase.update(Liquibase.java:179)
  at liquibase.integration.spring.SpringLiquibase.performUpdate(SpringLiquibase.java:353)
  at liquibase.integration.spring.SpringLiquibase.afterPropertiesSet(SpringLiquibase.java:305)
  at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.invokeInitMethods(AbstractAutowireCapableBeanFactory.java:1837)
  at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1774)
... 16 common frames omitted
Caused by: liquibase.exception.DatabaseException: java.sql.SQLFeatureNotSupportedException: setFetchSize
  at liquibase.snapshot.ResultSetCache.get(ResultSetCache.java:101)
  at liquibase.snapshot.JdbcDatabaseSnapshot$CachingDatabaseMetaData.getTables(JdbcDatabaseSnapshot.java:791)
  at liquibase.snapshot.jvm.TableSnapshotGenerator.snapshotObject(TableSnapshotGenerator.java:33)
  at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:66)
  at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
  at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:69)
  at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
  at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:69)
  at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
  at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:69)
  at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
  at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:69)
  at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
  at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:69)
  at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
  at liquibase.snapshot.jvm.JdbcSnapshotGenerator.snapshot(JdbcSnapshotGenerator.java:69)
  at liquibase.snapshot.SnapshotGeneratorChain.snapshot(SnapshotGeneratorChain.java:49)
  at liquibase.snapshot.DatabaseSnapshot.include(DatabaseSnapshot.java:286)
  at liquibase.snapshot.DatabaseSnapshot.init(DatabaseSnapshot.java:102)
  at liquibase.snapshot.DatabaseSnapshot.(DatabaseSnapshot.java:59)
  at liquibase.snapshot.JdbcDatabaseSnapshot.(JdbcDatabaseSnapshot.java:38)
  at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:217)
  at liquibase.snapshot.SnapshotGeneratorFactory.createSnapshot(SnapshotGeneratorFactory.java:246)
  at liquibase.snapshot.SnapshotGeneratorFactory.getDatabaseChangeLogTable(SnapshotGeneratorFactory.java:254)
  at liquibase.changelog.StandardChangeLogHistoryService.init(StandardChangeLogHistoryService.java:108)
... 23 common frames omitted
Caused by: java.sql.SQLFeatureNotSupportedException: setFetchSize
  at org.apache.shardingsphere.shardingjdbc.jdbc.unsupported.AbstractUnsupportedDatabaseMetaDataResultSet.setFetchSize(AbstractUnsupportedDatabaseMetaDataResultSet.java:123)
  at liquibase.snapshot.ResultSetCache$ResultSetExtractor.extract(ResultSetCache.java:293)
  at liquibase.snapshot.ResultSetCache$ResultSetExtractor.extract(ResultSetCache.java:288)
  at liquibase.snapshot.JdbcDatabaseSnapshot$CachingDatabaseMetaData$4.fastFetchQuery(JdbcDatabaseSnapshot.java:832)
  at liquibase.snapshot.ResultSetCache$SingleResultSetExtractor.fastFetch(ResultSetCache.java:347)
  at liquibase.snapshot.ResultSetCache.get(ResultSetCache.java:70)
... 47 common frames omitted


> You didn't fill in the bug report template, so I have to ask you here what's the release version? What's application(ShardingSphere-JDBC or ShardingSphere-Proxy)?
> 
  sharding-jdbc-spring-boot-starter version 4.1.1
  liquibase version 3.6.3

> When you use setFetchSize, what's the value you set?
> 
> You mentioned you haven met problem when use setFetchSize, what is the exact problem?
> 
  I didn’t use setFetchSize directly. I used setFetchSize internally in liquibase. I get an error when using both liquibase and Shardingsphere in the project.

> At last, if you have some findings and want to analyze, please do it in a separate section.

--- a/shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/resultset/DatabaseMetaDataResultSet.java
+++ b/shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/resultset/DatabaseMetaDataResultSet.java
@@ -59,11 +59,14 @@ public final class DatabaseMetaDataResultSet extends AbstractUnsupportedDatabase
     
     private final Iterator<DatabaseMetaDataObject> databaseMetaDataObjectIterator;
     
+    private final ResultSet resultSet;
+    
     private volatile boolean closed;
     
     private DatabaseMetaDataObject currentDatabaseMetaDataObject;
     
     public DatabaseMetaDataResultSet(final ResultSet resultSet, final Collection<ShardingSphereRule> rules) throws SQLException {
+        this.resultSet = resultSet;
         this.type = resultSet.getType();
         this.concurrency = resultSet.getConcurrency();
         this.rules = rules;
@@ -331,6 +334,26 @@ public final class DatabaseMetaDataResultSet extends AbstractUnsupportedDatabase
     public int getConcurrency() throws SQLException {
         checkClosed();
         return concurrency;
+    }
+    
+    @Override
+    public void setFetchDirection(final int direction) throws SQLException {
+        resultSet.setFetchDirection(direction);
+    }
+    
+    @Override
+    public int getFetchDirection() throws SQLException {
+        return resultSet.getFetchDirection();
+    }
+    
+    @Override
+    public int getFetchSize() throws SQLException {
+        return resultSet.getFetchSize();
+    }
+    
+    @Override
+    public void setFetchSize(final int rows) throws SQLException {
+        resultSet.setFetchSize(rows);
     }
     
     @Override

--- a/shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/resultset/DatabaseMetaDataResultSet.java
+++ b/shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/jdbc/core/resultset/DatabaseMetaDataResultSet.java
@@ -67,12 +67,12 @@ public final class DatabaseMetaDataResultSet extends AbstractUnsupportedDatabase
     
     public DatabaseMetaDataResultSet(final ResultSet resultSet, final Collection<ShardingSphereRule> rules) throws SQLException {
         this.resultSet = resultSet;
-        this.type = resultSet.getType();
-        this.concurrency = resultSet.getConcurrency();
         this.rules = rules;
-        this.resultSetMetaData = resultSet.getMetaData();
-        this.columnLabelIndexMap = initIndexMap();
-        this.databaseMetaDataObjectIterator = initIterator(resultSet);
+        type = resultSet.getType();
+        concurrency = resultSet.getConcurrency();
+        resultSetMetaData = resultSet.getMetaData();
+        columnLabelIndexMap = initIndexMap();
+        databaseMetaDataObjectIterator = initIterator(resultSet);
     }
     
     private Map<String, Integer> initIndexMap() throws SQLException {
@@ -379,7 +379,7 @@ public final class DatabaseMetaDataResultSet extends AbstractUnsupportedDatabase
         private final ArrayList<Object> objects;
         
         private DatabaseMetaDataObject(final int columnCount) {
-            this.objects = new ArrayList<>(columnCount);
+            objects = new ArrayList<>(columnCount);
         }
         
         public void addObject(final Object object) {

--- a/shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/jdbc/unsupported/AbstractUnsupportedDatabaseMetaDataResultSet.java
+++ b/shardingsphere-jdbc/shardingsphere-jdbc-core/src/main/java/org/apache/shardingsphere/driver/jdbc/unsupported/AbstractUnsupportedDatabaseMetaDataResultSet.java
@@ -109,26 +109,6 @@ public abstract class AbstractUnsupportedDatabaseMetaDataResultSet extends Abstr
     }
     
     @Override
-    public final void setFetchDirection(final int direction) throws SQLException {
-        throw new SQLFeatureNotSupportedException("setFetchDirection");
-    }
-    
-    @Override
-    public final int getFetchDirection() throws SQLException {
-        throw new SQLFeatureNotSupportedException("getFetchDirection");
-    }
-    
-    @Override
-    public final void setFetchSize(final int rows) throws SQLException {
-        throw new SQLFeatureNotSupportedException("setFetchSize");
-    }
-    
-    @Override
-    public final int getFetchSize() throws SQLException {
-        throw new SQLFeatureNotSupportedException("getFetchSize");
-    }
-    
-    @Override
     public final Statement getStatement() throws SQLException {
         throw new SQLFeatureNotSupportedException("getStatement");
     }

--- a/shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/core/resultset/DatabaseMetaDataResultSetTest.java
+++ b/shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/core/resultset/DatabaseMetaDataResultSetTest.java
@@ -104,6 +104,8 @@ public final class DatabaseMetaDataResultSetTest {
         when(result.getType()).thenReturn(ResultSet.TYPE_FORWARD_ONLY);
         when(result.getConcurrency()).thenReturn(ResultSet.CONCUR_READ_ONLY);
         when(result.next()).thenReturn(true, true, false);
+        when(result.getFetchDirection()).thenReturn(ResultSet.FETCH_FORWARD);
+        when(result.getFetchSize()).thenReturn(3);
         return result;
     }
     
@@ -335,6 +337,18 @@ public final class DatabaseMetaDataResultSetTest {
     @Test
     public void assertGetConcurrency() throws Exception {
         assertThat(databaseMetaDataResultSet.getConcurrency(), is(ResultSet.CONCUR_READ_ONLY));
+    }
+    
+    @Test
+    public void assertGetFetchDirection() throws Exception {
+        databaseMetaDataResultSet.setFetchDirection(ResultSet.FETCH_FORWARD);
+        assertThat(databaseMetaDataResultSet.getFetchDirection(), is(ResultSet.FETCH_FORWARD));
+    }
+    
+    @Test
+    public void assertGetFetchSize() throws Exception {
+        databaseMetaDataResultSet.setFetchSize(3);
+        assertThat(databaseMetaDataResultSet.getFetchSize(), is(3));
     }
     
     @Test(expected = SQLException.class)

--- a/shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/unsupported/UnSupportedDatabaseMetaDataResultSetTest.java
+++ b/shardingsphere-jdbc/shardingsphere-jdbc-core/src/test/java/org/apache/shardingsphere/driver/jdbc/unsupported/UnSupportedDatabaseMetaDataResultSetTest.java
@@ -112,26 +112,6 @@ public final class UnSupportedDatabaseMetaDataResultSetTest {
     }
     
     @Test(expected = SQLFeatureNotSupportedException.class)
-    public void assertSetFetchDirection() throws SQLException {
-        databaseMetaDataResultSet.setFetchDirection(ResultSet.FETCH_FORWARD);
-    }
-    
-    @Test(expected = SQLFeatureNotSupportedException.class)
-    public void assertGetFetchDirection() throws SQLException {
-        databaseMetaDataResultSet.getFetchDirection();
-    }
-    
-    @Test(expected = SQLFeatureNotSupportedException.class)
-    public void assertSetFetchSize() throws SQLException {
-        databaseMetaDataResultSet.setFetchSize(1);
-    }
-    
-    @Test(expected = SQLFeatureNotSupportedException.class)
-    public void assertGetFetchSize() throws SQLException {
-        databaseMetaDataResultSet.getFetchSize();
-    }
-    
-    @Test(expected = SQLFeatureNotSupportedException.class)
     public void assertGetStatement() throws SQLException {
         databaseMetaDataResultSet.getStatement();
     }

